### PR TITLE
Add ruff and use pyproject.toml (move away from setup.cfg)

### LINKED ISSUE #3684: Transition EvalML to use pyproject.toml only (move away from setup.cfg)
- pyproject.toml is the future of python package metadata and tool config
- "One file to rule them all"
- Examples
  - https://github.com/alteryx/featuretools/issues/2261
  - https://github.com/alteryx/woodwork/pull/1506

### PR DESCRIPTION
**Pull request by gsheni**
Fixes #3684

### CODE REVIEW

File: evalml/utils/cli_utils.py [hunk 1]
```diff
@@ -120,6 +120,9 @@ def standardize_format(packages, ignore_packages=None):
             continue
         name = CONDA_TO_PIP_NAME.get(package.name, package.name)
         if package.specs:
+            print(package)
+            print(package.specifier)
+            print(package.specs)
```
Comment (gsheni) [inline evalml/utils/cli_utils.py:125]:
```suggestion
```

**Review by cp2boston — APPROVED** [review]
looks good to me

File: Makefile [hunk 2]
```diff
@@ -10,19 +10,17 @@ clean:
...
 lint:
-	isort --check-only evalml
 	sh ./import_or_skip.sh
 	python docs/notebook_version_standardizer.py check-versions
 	python docs/notebook_version_standardizer.py check-execution
-	black evalml docs/source -t py39 --check
-	pydocstyle evalml/ --convention=google --add-ignore=D107 --add-select=D400 --match-dir='^(?!(tests)).*'
-	flake8 evalml
+	black evalml/ docs/source/ --check --config pyproject.toml
+	ruff evalml/ docs/source/
```
Comment (eccabay) [inline Makefile:17]: Have you been able to validate that this configuration catches all the style currently being enforced?

File: pyproject.toml [hunk 3]
```diff
@@ -1,6 +1,187 @@
...
+exclude_lines =[
+    "pragma: no cover",
+    "def __repr__",
+    "raise AssertionError",
+    "raise NotImplementedError",
+    "if __name__ == .__main__.:",
+    "if self._verbose:",
+    "if verbose:",
+    "if profile:",
+    "pytest.skip"
```
Comment (eccabay) [inline pyproject.toml:179]: Can we remove the unnecessary ones here? We do really want to be making sure we're testing things like verbosity and errors.

**Review by eccabay — CHANGES_REQUESTED** [review]
Looks good, just want to make sure we're not excluding too much from our code coverage

File: pyproject.toml (hunk 3 above)
Comment (gsheni) [reply pyproject.toml]: Done:
https://github.com/alteryx/evalml/pull/3928/commits/98aff8f20d4050b9f809f713bad61434e1ffd755#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711L172

File: Makefile (hunk 2 above)
Comment (gsheni) [reply Makefile:17]: Added pydocstyle check and fixes
https://github.com/alteryx/evalml/pull/3928/commits/98aff8f20d4050b9f809f713bad61434e1ffd755

File: evalml/tests/model_understanding_tests/test_partial_dependence.py [hunk 4]
```diff
@@ -2463,7 +2463,9 @@ def test_partial_dependence_jupyter_graph_check(
 def test_partial_dependence_fast_mode_after_dropped_feature(X_y_categorical_regression):
     """The feature selector component in this test will drop the 'time' feature, so confirm
     that fast mode handles the dropping of this feature correctly--namely that the feature
-    has no impact on predictions"""
+    has no impact on predictions
+    .
+    """
```
Comment (eccabay) [inline evalml/tests/model_understanding_tests/test_partial_dependence.py:2468]: This formatting doesn't look quite right

**Review by eccabay — APPROVED** [review]

File: evalml/tests/model_understanding_tests/test_partial_dependence.py (hunk 4 above)
Comment (gsheni) [reply evalml/tests/model_understanding_tests/test_partial_dependence.py]: Fixed:
https://github.com/alteryx/evalml/pull/3928/commits/2630378a3722ceac75516034279362527eff5412

**Pull request merged**